be/src/storage/level_compaction_policy.cpp

```diff
@@ -0,0 +1,285 @@
+    uint32_t input_rows_num = 0;
+    size_t input_size = 0;
+    // add the base rowset to input_rowsets
+    Rowset* base_rowset = *_compaction_context->rowset_levels[2].begin();
+    rowsets->push_back(base_rowset->shared_from_this());
+    input_rows_num += base_rowset->num_rows();
+    input_size += base_rowset->data_disk_size();
+    // add level-1 rowsets
+    for (auto rowset : _compaction_context->rowset_levels[1]) {
+        rowsets->push_back(rowset->shared_from_this());
+        input_rows_num += rowset->num_rows();
+        input_size += rowset->data_disk_size();
+    }
+}
+
+std::shared_ptr<CompactionTask> LevelCompactionPolicy::_create_level_1_compaction() {
+    std::vector<RowsetSharedPtr> input_rowsets;
+    _pick_level_1_rowsets(&input_rowsets);
+    if (input_rowsets.size() <= 1) {
+        LOG(INFO) << "no suitable version for compaction. size:" << input_rowsets.size();
+        return nullptr;
+    }
+
+    if (input_rowsets.size() == 2 && input_rowsets[0]->empty()) {
+        // the tablet is with rowset: [0-x], [x+1-y], and [0-x] is empty,
+        // no need to do base compaction.
+        LOG(INFO) << "only two rowset. one is [0-x] and it is empty. do not compact. tablet:"
+                  << _compaction_context->tablet->tablet_id();
+        return nullptr;
+    }
+    DCHECK(input_rowsets.size() > 0) << "input rowsets size can not be empty";
+    auto st = _check_version_continuity(input_rowsets);
+    if (!st.ok()) {
+        return nullptr;
+    }
+
+    Version output_version;
+    output_version.first = (*input_rowsets.begin())->start_version();
+    output_version.second = (*input_rowsets.rbegin())->end_version();
+
+    CompactionTaskFactory factory(output_version, _compaction_context->tablet, std::move(input_rowsets),
+                                  _compaction_context->get_score(), 1);
+    std::shared_ptr<CompactionTask> compaction_task = factory.create_compaction_task();
+    return compaction_task;
+}
+
+double LevelCompactionPolicy::_get_level_0_compaction_score() {
+    uint32_t segment_num_score = 0;
+    size_t rowsets_size = 0;
+    for (auto& rowset : _compaction_context->rowset_levels[0]) {
+        segment_num_score += rowset->rowset_meta()->get_compaction_score();
+        rowsets_size += rowset->rowset_meta()->total_disk_size();
+    }
+
+    double num_score = static_cast<double>(segment_num_score) / config::min_cumulative_compaction_num_singleton_deltas;
+    double size_score = static_cast<double>(rowsets_size) / config::min_level_0_compaction_size;
+    double score = std::max(num_score, size_score);
+    VLOG(2) << "tablet:" << _compaction_context->tablet->tablet_id() << ", compaction score:" << score
+            << ", size_score:" << size_score << ", num_score:" << num_score << ", rowsets_size:" << rowsets_size;
+    return score;
+}
+
+double LevelCompactionPolicy::_get_level_1_compaction_score() {
+    uint32_t segment_num_score = 0;
+    size_t level_1_rowsets_size = 0;
```

Comment:
I think should consider the possibility of starve
ex. some tablet with huge write throughput, score may always larger than 1
some tablet with large level-1 comparative to level-2 , and it's score always nearly 1
the long time no compaction tablet is the same situation